router/core/src/main/scala/com/twitter/finagle/buoyant/DstBindingFactory.scala

```diff
@@ -126,12 +133,16 @@ object DstBindingFactory {
     boundMk: Mk[Dst.Bound, Req, Rsp] = Mk.identity[Dst.Bound, Req, Rsp],
     namer: NameInterpreter = DefaultInterpreter,
     statsReceiver: StatsReceiver = DefaultStatsReceiver,
-    capacity: Capacity = Capacity.default
-  ) extends DstBindingFactory[Req, Rsp] {
+    capacity: Capacity = Capacity.default,
+    bindingTimeout: BindingTimeout = BindingTimeout.default
+  )(implicit timer: Timer = DefaultTimer.twitter) extends DstBindingFactory[Req, Rsp] {
     private[this]type Cache[Key] = XXX_ServiceFactoryCache[Key, Req, Rsp]
 
     def apply(dst: Dst, conn: ClientConnection): Future[Service[Req, Rsp]] = dst match {
-      case path: Dst.Path => pathCache(path, conn)
+      case path: Dst.Path => pathCache(path, conn).raiseWithin(
+        bindingTimeout.timeout,
+        new TimeoutException(s"Binding ${path.path.show} exceeded ${bindingTimeout.timeout}")
+      )
```

Comment:
should we special case the max timeout so we don't always set timeouts?

i.e.
```scala
val f = pathCache(path, conn)
bindingTimeout match {
  case BindingTimeout(Duration.Top) => f
  case BindingTimeout(timeout) => f.raiseWithin(timeout, exception)
}
```

I think we could also val this exception -- it doesn't need to be instantiated on each request but can just be a private on the class.